#!/applications/R/R-3.4.0/bin/Rscript
### Perform differential expression analysis using read counts generated by Salmon

# R version 3.4.0 (invoke this version by specifying path at command line:
# /applications/R/R-3.4.0/bin/R , or if running as a script:
# /applications/R/R-3.4.0/bin/Rscript)
# DESeq2 version 1.16.1
# Note that this R version or later is required for DESeq2 version 1.16 or later,
# in which "the log2 fold change shrinkage is no longer default for the DESeq and
# nbinomWaldTest functions". DESeq2 version 1.16 introduces "a separate function
# lfcShrink, which performs log2 fold change shrinkage for visualization and ranking
# of genes." (see https://support.bioconductor.org/p/95695/ and
# https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#changes)

# Code below based on:
# https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html

# Usage on node7 with Condor submission system
# csmit -m 1G -c 24 "tximport_DESeq2.R ../samples.txt DESeq2_analysis_01" 

sampleFile <- "/projects/ajt200/BAM_masters/RNAseq_meiocyte_Walker_Feng_2018_NatGenet/WT/samples.txt"
analysisDir <- "DESeq2_analysis_01"

args <- commandArgs(trailingOnly = TRUE)
sampleFile <- args[1]
analysisDir <- args[2]

outDir <- paste0(dirname(sampleFile), "/", analysisDir, "/")

## Import transcript abundance datasets with tximport package

library(tximport)
print(packageVersion("tximport"))
#[1] ‘1.4.0’

# Read in table of sample IDs that will be used to specify paths to count files
samples <- read.table(sampleFile, header = T)
print(samples)

# Specify paths to count files
files <- file.path(dirname(sampleFile),
                   samples$tissue, samples$directory, samples$sample,
                   "quant.sf")
# Set 1:#_samples
names(files) <- paste0("sample", 1:6)
all(file.exists(files))
#[1] TRUE

# Create a dataframe of transcript IDs and corresponding gene IDs
transID <- read.table(files[1], colClasses = c(NA, rep("NULL", 4)), header = T)
tx2gene <- data.frame(cbind(as.vector(transID[,1]), substr(transID[,1], 1, 9)))
colnames(tx2gene) <- c("TXNAME", "GENEID")

# Import transcript-level counts, summarised at gene level
# (reads that map to transcript IDs with a common parent gene ID are pooled)
library(readr)
txi <- tximport(files, type = "salmon", tx2gene = tx2gene)
print(names(txi))
#[1] "abundance"           "counts"              "length"             
#[4] "countsFromAbundance"

txi_scaled <- tximport(files, type = "salmon", tx2gene = tx2gene,
                       countsFromAbundance = "scaledTPM")
txi_lengthScaled <- tximport(files, type = "salmon", tx2gene = tx2gene,
                             countsFromAbundance = "lengthScaledTPM")

txi_counts <- txi$counts
txi_tpm <- txi$abundance
colnames(txi_counts) <- c("leaf_Rep1", "leaf_Rep2", "leaf_Rep3",
                          "meiocyte_Rep1", "meiocyte_Rep2", "meiocyte_Rep3")
colnames(txi_tpm) <- c("leaf_Rep1", "leaf_Rep2", "leaf_Rep3",
                       "meiocyte_Rep1", "meiocyte_Rep2", "meiocyte_Rep3")
write.table(as.data.frame(txi_counts),
            file = paste0(outDir, "salmon_counts_WT_RNAseq_leaf_and_meiocyte.txt"),
            sep = "\t", quote = F)
write.table(as.data.frame(txi_tpm),
            file = paste0(outDir, "salmon_TPM_WT_RNAseq_leaf_and_meiocyte.txt"),
            sep = "\t", quote = F)


