#!/applications/R/R-3.4.0/bin/Rscript
### Perform differential expression analysis using read counts generated by Salmon

# R version 3.4.0 (invoke this version by specifying path at command line:
# /applications/R/R-3.4.0/bin/R , or if running as a script:
# /applications/R/R-3.4.0/bin/Rscript)
# DESeq2 version 1.16.1
# Note that this R version or later is required for DESeq2 version 1.16 or later,
# in which "the log2 fold change shrinkage is no longer default for the DESeq and
# nbinomWaldTest functions". DESeq2 version 1.16 introduces "a separate function
# lfcShrink, which performs log2 fold change shrinkage for visualization and ranking
# of genes." (see https://support.bioconductor.org/p/95695/ and
# https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#changes)

# Code below based on:
# https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html

# Usage on node7 with Condor submission system
# csmit -m 1G -c 24 "tximport_DESeq2.R ../samples.txt DESeq2_analysis_01 0.1"

sampleFile <- "/projects/ajt200/BAM_masters/RNAseq_meiocyte_Walker_Feng_2018_NatGenet/WT/samples.txt"

args <- commandArgs(trailingOnly = TRUE)
sampleFile <- args[1]
analysisDir <- args[2]
FDR <- args[3]

outDir <- paste0(dirname(sampleFile), "/", analysisDir, "/")
plotDir <- paste0(dirname(sampleFile), "/", analysisDir, "/plots/")
downReg_plotDir <- paste0(plotDir, "DEGdownReg_counts/")
upReg_plotDir <- paste0(plotDir, "DEGupReg_counts/")
system(paste0("[ -d ", plotDir, " ] || mkdir ", plotDir))
system(paste0("[ -d ", downReg_plotDir, " ] || mkdir ", downReg_plotDir))
system(paste0("[ -d ", upReg_plotDir, " ] || mkdir ", upReg_plotDir))

## Import transcript abundance datasets with tximport package

library(tximport)
print(packageVersion("tximport"))
#[1] ‘1.4.0’

# Read in table of sample IDs that will be used to specify paths to count files
samples <- read.table(sampleFile, header = T)
print(samples)

# Specify paths to count files
files <- file.path(dirname(sampleFile),
                   samples$tissue, samples$directory, samples$sample,
                   "quant.sf")
# Set 1:#_samples
names(files) <- paste0("sample", 1:6)
all(file.exists(files))
#[1] TRUE

# Create a dataframe of transcript IDs and corresponding gene IDs
transID <- read.table(files[1], colClasses = c(NA, rep("NULL", 4)), header = T)
tx2gene <- data.frame(cbind(as.vector(transID[,1]), substr(transID[,1], 1, 9)))
colnames(tx2gene) <- c("TXNAME", "GENEID")

# Import transcript-level counts, summarised at gene level
# (reads that map to transcript IDs with a common parent gene ID are pooled)
library(readr)
txi <- tximport(files, type = "salmon", tx2gene = tx2gene)
print(names(txi))
#[1] "abundance"           "counts"              "length"             
#[4] "countsFromAbundance"

# Import transcript-level counts, summarised at transcript level with "txOut = TRUE"
txi.tx <- tximport(files, type = "salmon", txOut = TRUE, tx2gene = tx2gene)
# Then summarise to gene level
txi.sum <- summarizeToGene(txi.tx, tx2gene)
# These two approaches should produce identical results
print(all.equal(txi$counts, txi.sum$counts))
#[1] TRUE

print(head(txi$counts))
#           sample1 sample2 sample3 sample4 sample5 sample6
#AT1G01010  29.0000      22      42       7       0      34
#AT1G01020  62.0000      32      92      23       5      15
#AT1G01030  40.0000      14      21      20       0       8
#AT1G01040 587.0000     784     163      98      40     187
#AT1G01050 169.8244     135     153     153      96      40
#AT1G01060  14.0000      14      51      17      10       8


## Use with downstream Bioconductor differential expression package
## DESeq2: http://www.bioconductor.org/help/workflows/rnaseqGene/

library(DESeq2)
print(packageVersion("DESeq2"))
#[1] ‘1.16.1’
sampleTable <- data.frame(sample = c("Col-0 leaf RNA-seq Rep1", "Col-0 leaf RNA-seq Rep2", "Col-0 leaf RNA-seq Rep3",
                                     "Col-0 meiocyte RNA-seq Rep1", "Col-0 meiocyte RNA-seq Rep2", "Col-0 meiocyte RNA-seq Rep3"),
                          condition = factor(rep.int(c("leaf", "meiocyte"),
                                                     times = c(3, 3))))
rownames(sampleTable) <- colnames(txi$counts)
print(sampleTable)
#                             sample condition
#sample1     Col-0 leaf RNA-seq Rep1      leaf
#sample2     Col-0 leaf RNA-seq Rep2      leaf
#sample3     Col-0 leaf RNA-seq Rep3      leaf
#sample4 Col-0 meiocyte RNA-seq Rep1  meiocyte
#sample5 Col-0 meiocyte RNA-seq Rep2  meiocyte
#sample6 Col-0 meiocyte RNA-seq Rep3  meiocyte


## Differential expression analysis

# Unless there is high within-group variability,
# the DESeq developers recommend that the DESeq() function is run to fit a model
# using samples from all groups, followed by use of the "contrast" argument of the
# results() function to make comparisons between two groups.
# "The model fit by DESeq estimates a single dispersion parameter for each gene,
# which defines how far we expect the observed count for a sample will be from the
# mean value from the model given its size factor and its condition group."
# (https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)

## Col-0 meiocyte vs Col-0 leaf

# Re-import DESeqDataSet to redefine "dds"
dds <- DESeqDataSetFromTximport(txi = txi, colData = sampleTable, design = ~condition)

# Pre-filter the datasets
print(nrow(dds))
#[1] 27586
# Retain only rows that have more than a single count across all samples
dds <- dds[rowSums(counts(dds)) > 1,]
print(nrow(dds))
#[1] 24574

# Contrast: Col-0 meiocyte vs Col-0 leaf 
dds <- DESeq(dds)
res_meiocyteVleaf <- results(dds,
                             contrast = c("condition", "meiocyte", "leaf"),
                             alpha = as.numeric(FDR))
print(mcols(res_meiocyteVleaf, use.names = T))
print("summary of res_meiocyteVleaf")
print(summary(res_meiocyteVleaf))
save(res_meiocyteVleaf,
     file = paste0(outDir, "res_meiocyteVleaf_", as.character(FDR), ".RData"))

# MA-plots
# These show the log2(fold changes) in gene expression values for a variable (e.g., treated vs untreated)
# over the mean of normalized counts for all the samples in the DESeqDataSet

# DESeq2 provides for use of a "Bayesian procedure to moderate (or "shrink") log2 fold changes from genes
# with very low counts and highly variable counts"
# Before generating an MA-plot, the lfcShrink() function should be used to moderate the log2(fold changes) for the contrast
res_meiocyteVleaf_lfcShrink <- lfcShrink(dds, contrast = c("condition", "meiocyte", "leaf"), res = res_meiocyteVleaf)
pdf(paste0(plotDir, "MAplot_res_meiocyteVleaf_", as.character(FDR), "_lfcShrink.pdf"), height = 10, width = 6)
par(mfrow = c(2, 1))
par(mar = c(4.1, 4.1, 2.1, 2.1))
plotMA(res_meiocyteVleaf_lfcShrink, ylim = c(-14, 14),
       xlab = "Mean of normalized counts",
       ylab = "log2(fold change)")
plotMA(res_meiocyteVleaf_lfcShrink, ylim = c(-5, 5),
       xlab = "Mean of normalized counts",
       ylab = "log2(fold change)")
dev.off()

# Histograms of unadjusted and FDR-adjusted p-values
pdf(paste0(plotDir, "hist_res_meiocyteVleaf_", as.character(FDR), "_pvals_unadjusted.pdf"))
hist(res_meiocyteVleaf$pvalue[res_meiocyteVleaf$baseMean > 1], breaks = 0:20/20,
     col = "grey50", border = "white")
dev.off()

pdf(paste0(plotDir, "hist_res_meiocyteVleaf_", as.character(FDR), "_pvals_BH_FDR_adjusted.pdf"))
hist(res_meiocyteVleaf$padj[res_meiocyteVleaf$baseMean > 1], breaks = 0:20/20,
     col = "grey50", border = "white")
dev.off()

# Independent filtering
# Create bins
qs <- c(0, quantile(res_meiocyteVleaf$baseMean[res_meiocyteVleaf$baseMean > 0], 0:6/6))
# Bin genes by base mean (mean normalized count) using cut
bins <- cut(res_meiocyteVleaf$baseMean, qs)
# Rename the levels of the bins using the middle point
levels(bins) <- paste0("~", round(signif((qs[-1] + qs[-length(qs)])/2, 2)))
fractionSig <- tapply(res_meiocyteVleaf$pvalue, bins, function(p)
                 mean(p < 0.05, na.rm = TRUE))
pdf(paste0(plotDir, "barplot_res_meiocyteVleaf_", as.character(FDR), "_fraction_small_pval_genes.pdf"))
barplot(fractionSig, xlab = "Mean normalized count",
                     ylab = "Fraction of small P values")
dev.off()


## Exporting results

res_meiocyteVleaf_lfcShrink_Chr <- res_meiocyteVleaf_lfcShrink[!grepl("ATM", rownames(res_meiocyteVleaf_lfcShrink)) &
                                                               !grepl("ATC", rownames(res_meiocyteVleaf_lfcShrink)), ]

# Subset results table to significant down-regulated genes on chromosomes in taf4b
res_meiocyteVleaf_lfcShrink_Chr_Sig_downReg <- subset(res_meiocyteVleaf_lfcShrink_Chr, padj < as.numeric(FDR) & log2FoldChange < 0)
# Sort by increasing log2 fold change estimate
# (Genes with strongest down-regulation)
res_meiocyteVleaf_lfcShrink_Chr_Sig_downRegSorted <- res_meiocyteVleaf_lfcShrink_Chr_Sig_downReg[order(res_meiocyteVleaf_lfcShrink_Chr_Sig_downReg$log2FoldChange),]
# Subset results table to significant up-regulated genes on chromosomes in taf4b
res_meiocyteVleaf_lfcShrink_Chr_Sig_upReg <- subset(res_meiocyteVleaf_lfcShrink_Chr, padj < as.numeric(FDR) & log2FoldChange > 0)
# Sort by decreasing log2(fold change) estimate
# (Genes with strongest up-regulation) 
res_meiocyteVleaf_lfcShrink_Chr_Sig_upRegSorted <- res_meiocyteVleaf_lfcShrink_Chr_Sig_upReg[order(res_meiocyteVleaf_lfcShrink_Chr_Sig_upReg$log2FoldChange, decreasing = T),]

# Convert DataFrame objects (IRanges package) to data.frame objects that can be processed by write.table
res_meiocyteVleaf_lfcShrink_Chr_Sig_downRegSortedDF <- as.data.frame(res_meiocyteVleaf_lfcShrink_Chr_Sig_downRegSorted)
res_meiocyteVleaf_lfcShrink_Chr_Sig_upRegSortedDF <- as.data.frame(res_meiocyteVleaf_lfcShrink_Chr_Sig_upRegSorted)
write.table(res_meiocyteVleaf_lfcShrink_Chr_Sig_downRegSortedDF,
            file = paste0(outDir, "res_meiocyteVleaf_", as.character(FDR), "_lfcShrink_Chr_Sig", as.character(FDR), "_downRegSortedDF.txt"))
write.table(res_meiocyteVleaf_lfcShrink_Chr_Sig_upRegSortedDF,
            file = paste0(outDir, "res_meiocyteVleaf_", as.character(FDR), "_lfcShrink_Chr_Sig", as.character(FDR), "_upRegSortedDF.txt"))

res_meiocyteVleaf_lfcShrink_Chr_Sig_downRegGeneIDs <- rownames(res_meiocyteVleaf_lfcShrink_Chr_Sig_downRegSortedDF)
res_meiocyteVleaf_lfcShrink_Chr_Sig_upRegGeneIDs <- rownames(res_meiocyteVleaf_lfcShrink_Chr_Sig_upRegSortedDF)
write.table(res_meiocyteVleaf_lfcShrink_Chr_Sig_downRegGeneIDs,
            file = paste0(outDir, "res_meiocyteVleaf_", as.character(FDR), "_lfcShrink_Chr_Sig", as.character(FDR), "_downRegGeneIDs.txt"))
write.table(res_meiocyteVleaf_lfcShrink_Chr_Sig_upRegGeneIDs,
            file = paste0(outDir, "res_meiocyteVleaf_", as.character(FDR), "_lfcShrink_Chr_Sig", as.character(FDR), "_upRegGeneIDs.txt"))


## Generate plots of log-transformed normalized counts at differentially expressed genes in meiocyte vs leaf tissue 

library(ggplot2)
library(ggbeeswarm)
library(parallel)
library(org.At.tair.db)
TAIRsymbol <- org.At.tairSYMBOL
# Get TAIR gene identifiers that are mapped to a gene symbol
mapped_geneIDs <- mappedkeys(TAIRsymbol)

dds <- DESeqDataSetFromTximport(txi = txi, colData = sampleTable, design = ~condition)

meiocyteVleaf_lfcShrink_Chr_Sig_downRegGeneIDs <- as.character(read.table(file = paste0(outDir, "res_meiocyteVleaf_", as.character(FDR), "_lfcShrink_Chr_Sig", as.character(FDR), "_downRegGeneIDs.txt"))$x)
meiocyteVleaf_lfcShrink_Chr_Sig_upRegGeneIDs <- as.character(read.table(file = paste0(outDir, "res_meiocyteVleaf_", as.character(FDR), "_lfcShrink_Chr_Sig", as.character(FDR), "_upRegGeneIDs.txt"))$x)

# Count-plotting function
countPlotFun <- function(geneID, contrast, plotDir) {
  geneCounts <- plotCounts(dds, gene = geneID, intgroup = c("condition", "sample"),
                           normalized = T, transform = T, returnData = T)
  geneCounts$condition <- factor(geneCounts$condition,
                                 levels = c("leaf", "meiocyte"))
  geneCountsPlot <- ggplot(geneCounts,
                           aes(x = condition, y = count, colour = sample)) +
                      scale_y_log10(breaks = pretty(geneCounts$count)) +
                      geom_beeswarm(cex = 3) +
                      xlab("Tissue") +
                      ylab("log-transformed normalized count") +
                      labs(colour = "Sample") +
                      theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
                      theme_classic() +
                      theme(panel.border = element_rect(colour = "black",
                                                        fill = NA, size = 1)) +
                      ggtitle(paste0(geneID,
                                     " (", TAIRsymbol[[geneID]][1], ")")) +
                      theme(plot.title = element_text(hjust = 0.5))
  ggsave(geneCountsPlot,
         file = paste0(plotDir, contrast, "_", geneID, "_",
                       gsub("/", "|",
                       TAIRsymbol[[geneID]][1]), "_normalized_counts.pdf"),
         width = 16, height = 10, units = "cm")
}

mclapply(seq_along(meiocyteVleaf_lfcShrink_Chr_Sig_downRegGeneIDs), function(x) {
  countPlotFun(geneID = meiocyteVleaf_lfcShrink_Chr_Sig_downRegGeneIDs[x],
               contrast = paste0("meiocyteVleaf_", as.character(FDR),
                                 "_lfcShrink_Chr_Sig", as.character(FDR),
                                 "_downReg"),
               plotDir = downReg_plotDir)
}, mc.cores = 24)

mclapply(seq_along(meiocyteVleaf_lfcShrink_Chr_Sig_upRegGeneIDs), function(x) {
  countPlotFun(geneID = meiocyteVleaf_lfcShrink_Chr_Sig_upRegGeneIDs[x],
               contrast = paste0("meiocyteVleaf_", as.character(FDR),
                                 "_lfcShrink_Chr_Sig", as.character(FDR),
                                 "_upReg"),
               plotDir = upReg_plotDir)
}, mc.cores = 24)


