#!/applications/R/R-3.4.0/bin/Rscript
### Perform differential expression analysis using read counts generated by Salmon

# R version 3.4.0 (invoke this version by specifying path at command line: /applications/R/R-3.4.0/bin/R , or if running as a script: /applications/R/R-3.4.0/bin/Rscript)
# DESeq2 version 1.16.1
# Note that this R version or later is required for DESeq2 version 1.16 or later,
# in which "the log2 fold change shrinkage is no longer default for the DESeq and nbinomWaldTest functions".
# DESeq2 version 1.16 introduces "a separate function lfcShrink, which performs log2 fold change shrinkage
# for visualization and ranking of genes." (see https://support.bioconductor.org/p/95695/ and
# https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#changes)

## Import transcript abundance datasets with tximport package
# Code below based on https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html

# Usage: tximport_sample_exploration.R ../samples.txt DESeq2_analysis_01

args <- commandArgs(trailingOnly = TRUE)
sampleFile <- args[1]
analysisDir <- args[2]

outDir <- paste0(dirname(sampleFile), "/", analysisDir, "/")
plotDir <- paste0(dirname(sampleFile), "/", analysisDir, "/sample_exploration/")
system(paste0("[ -d ", plotDir, " ] || mkdir ", plotDir))

library(tximport)
print(packageVersion("tximport"))
#[1] ‘1.4.0’

# Read in table of sample IDs that will be used to specify paths to count files
samples <- read.table(sampleFile, header = T)
print(samples)

# Specify paths to count files
files <- file.path(dirname(sampleFile), samples$tissue, samples$directory, samples$sample, "quant.sf")
# Set 1:#_samples
names(files) <- paste0("sample", 1:6)
all(file.exists(files))
#[1] TRUE

# Create a dataframe of transcript IDs and corresponding gene IDs
transID <- read.table(files[1], colClasses = c(NA, rep("NULL", 4)), header = T)
tx2gene <- data.frame(cbind(as.vector(transID[,1]), substr(transID[,1], 1, 9)))
colnames(tx2gene) <- c("TXNAME", "GENEID")

# Import transcript-level counts, summarised at gene level
# (reads that map to transcript IDs with a common parent gene ID are pooled)
library(readr)
txi <- tximport(files, type = "salmon", tx2gene = tx2gene)
print(names(txi))
#[1] "abundance"           "counts"              "length"             
#[4] "countsFromAbundance"

# Import transcript-level counts, summarised at transcript level with "txOut = TRUE"
txi.tx <- tximport(files, type = "salmon", txOut = TRUE, tx2gene = tx2gene)
# Then summarise to gene level
txi.sum <- summarizeToGene(txi.tx, tx2gene)
# These two approaches should produce identical results
print(all.equal(txi$counts, txi.sum$counts))
#[1] TRUE

print(head(txi$counts))


## Use with downstream Bioconductor differential expression package
## DESeq2: http://www.bioconductor.org/help/workflows/rnaseqGene/

library(DESeq2)
print(packageVersion("DESeq2"))
#[1] ‘1.16.1’
sampleTable <- data.frame(sample = c("Col-0 leaf RNA-seq Rep1", "Col-0 leaf RNA-seq Rep2", "Col-0 leaf RNA-seq Rep3",
                                     "Col-0 meiocyte RNA-seq Rep1", "Col-0 meiocyte RNA-seq Rep2", "Col-0 meiocyte RNA-seq Rep3"),
                          condition = factor(rep.int(c("leaf", "meiocyte"), times = c(3, 3))))
rownames(sampleTable) <- colnames(txi$counts)
print(sampleTable)

dds <- DESeqDataSetFromTximport(txi = txi, colData = sampleTable, design = ~condition)

# Pre-filter the dataset
print(nrow(dds))
#[1] 27586
# Retain only rows that have more than a single count across all samples
dds <- dds[rowSums(counts(dds)) > 1,]
print(nrow(dds))
#[1] 24574


## The rlog and variance stabilizing transformations
# see http://www.bioconductor.org/help/workflows/rnaseqGene/#the-rlog-and-variance-stabilizing-transformations

rld <- rlog(dds, blind = FALSE)
print(head(assay(rld), 3))

vsd <- vst(dds, blind = FALSE)
print(head(assay(vsd), 3))

# Visualise the effect of transformation
library(dplyr)
library(ggplot2)
library(hexbin)

# For the log2 approach, estimate size factors to account for sequencing depth
# Sequencing-depth correction is done automatically for rlog and vst
dds <- estimateSizeFactors(dds)

df <- bind_rows(
  as_data_frame(log2(counts(dds, normalized = T)[,1:2]+1)) %>%
    mutate(transformation = "log2(normalized counts + 1)"),
  as_data_frame(assay(rld)[,1:2]) %>% mutate(transformation = "rlog"),
  as_data_frame(assay(vsd)[,1:2]) %>% mutate(transformation = "vst"))

colnames(df)[1:2] <- c("Sample 1 transformed counts", "Sample 2 transformed counts")

plot_transformed_counts <- ggplot(df, aes(x = `Sample 1 transformed counts`, y = `Sample 2 transformed counts`)) +
                             geom_hex(bins = 80) + coord_fixed() + facet_grid(. ~ transformation) +
                             labs(fill = "Occurrences") +
                             theme_classic()
                             #theme(panel.border = element_rect(colour = "black", fill = NA, size = 1))
ggsave(plot_transformed_counts,
       file = paste0(plotDir, "Sample1_vs_Sample2_transformed_counts_log2countsPlus1_rlog_vst.pdf"))


## Sample distances

# Sample distances using the rlog-transformed counts
rld_sampleDists <- dist(t(assay(rld)))
print(rld_sampleDists)

library(pheatmap)
library(RColorBrewer)

# Heatmap of sample distances using the rlog-transformed counts
rld_sampleDistMatrix <- as.matrix(rld_sampleDists)
rownames(rld_sampleDistMatrix) <- c("Col-0 leaf RNA-seq Rep1", "Col-0 leaf RNA-seq Rep2", "Col-0 leaf RNA-seq Rep3",
                                    "Col-0 meiocyte RNA-seq Rep1", "Col-0 meiocyte RNA-seq Rep2", "Col-0 meiocyte RNA-seq Rep3")
colnames(rld_sampleDistMatrix) <- NULL
mycols <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)
pdf(paste0(plotDir, "sample_distances_heatmap_rlog.pdf"), height = 5, width = 7.5, onefile = F)
pheatmap(rld_sampleDistMatrix,
         clustering_distance_rows = rld_sampleDists,
         clustering_distance_cols = rld_sampleDists,
         col = mycols)
dev.off()

# Sample distances using the vst-transformed counts
vsd_sampleDists <- dist(t(assay(vsd)))
print(vsd_sampleDists)

library(pheatmap)
library(RColorBrewer)

# Heatmap of sample distances using the vst-transformed counts
vsd_sampleDistMatrix <- as.matrix(vsd_sampleDists)
rownames(vsd_sampleDistMatrix) <- c("Col-0 leaf RNA-seq Rep1", "Col-0 leaf RNA-seq Rep2", "Col-0 leaf RNA-seq Rep3",
                                    "Col-0 meiocyte RNA-seq Rep1", "Col-0 meiocyte RNA-seq Rep2", "Col-0 meiocyte RNA-seq Rep3")
colnames(vsd_sampleDistMatrix) <- NULL
mycols <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)
pdf(paste0(plotDir, "sample_distances_heatmap_vst.pdf"), height = 5, width = 7.5, onefile = F)
pheatmap(vsd_sampleDistMatrix,
         clustering_distance_rows = vsd_sampleDists,
         clustering_distance_cols = vsd_sampleDists,
         col = mycols)
dev.off()

# Sample distances using the Poisson Distance
library(PoiClaClu)
poisd <- PoissonDistance(t(counts(dds)))
print(poisd)
#Value of alpha used to transform data:  0.3130612
#This type of normalization was performed: mle
#Dissimilarity computed for  6  observations.

# Heatmap
samplePoisDistMatrix <- as.matrix(poisd$dd)
rownames(samplePoisDistMatrix) <- c("Col-0 leaf RNA-seq Rep1", "Col-0 leaf RNA-seq Rep2", "Col-0 leaf RNA-seq Rep3",
                                    "Col-0 meiocyte RNA-seq Rep1", "Col-0 meiocyte RNA-seq Rep2", "Col-0 meiocyte RNA-seq Rep3")
colnames(samplePoisDistMatrix) <- NULL
pdf(paste0(plotDir, "sample_distances_heatmap_Poisson.pdf"), height = 5, width = 7.5, onefile = F)
pheatmap(samplePoisDistMatrix,
         clustering_distance_rows = poisd$dd,
         clustering_distance_cols = poisd$dd,
         col = mycols)
dev.off()

# PCA plots for visualising sample-to-sample distances
# rlog-transformed counts
PCAplot_rlog_data <- DESeq2::plotPCA(rld, intgroup = c("condition", "sample"), returnData = T)
PCAplot_rlog_data
# Obtain percentage variance explained by PC1 and PC2 for plotting using ggplot2
percentVar <- round(100 * attr(PCAplot_rlog_data, "percentVar"))

PCAggplot_rlog <- ggplot(PCAplot_rlog_data, aes(x = PC1, y = PC2,
                                                shape = condition,
                                                colour = sample)) +
                    geom_point(size = 3) +
                    xlab(paste0("PC1: ", percentVar[1], "% variance")) +
                    ylab(paste0("PC2: ", percentVar[2], "% variance")) +
                    theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
                    theme_classic() +
                    theme(panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
                    coord_fixed()
ggsave(PCAggplot_rlog,
       file = paste0(plotDir, "PCAggplot_rlog.pdf"), height = 20, width = 20, units = "cm")

# vst-transformed counts
PCAplot_vst_data <- DESeq2::plotPCA(vsd, intgroup = c("condition", "sample"), returnData = T)
PCAplot_vst_data
# Obtain percentage variance explained by PC1 and PC2 for plotting using ggplot2
percentVar <- round(100 * attr(PCAplot_vst_data, "percentVar"))

PCAggplot_vst <- ggplot(PCAplot_vst_data, aes(x = PC1, y = PC2,
                                                shape = condition,
                                                colour = sample)) +
                    geom_point(size = 3) +
                    xlab(paste0("PC1: ", percentVar[1], "% variance")) +
                    ylab(paste0("PC2: ", percentVar[2], "% variance")) +
                    theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
                    theme_classic() +
                    theme(panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
                    coord_fixed()
ggsave(PCAggplot_vst,
       file = paste0(plotDir, "PCAggplot_vst.pdf"), height = 20, width = 20, units = "cm")

# MDS (multi-dimensional scaling) plots for visualising sample-to-sample distances
# rlog-transformed counts
mds_rlog <- as.data.frame(colData(rld)) %>%
         cbind(cmdscale(rld_sampleDistMatrix))
MDSplot_rlog <- ggplot(mds_rlog, aes(x = `1`, y = `2`,
                                shape = condition,
                                colour = sample)) +
                  geom_point(size = 3) +
                  theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
                  theme_classic() +
                  theme(panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
                  coord_fixed()
ggsave(MDSplot_rlog,
       file = paste0(plotDir, "MDSplot_rlog.pdf"), height = 20, width = 20, units = "cm")

# vst-transformed counts
mds_vst <- as.data.frame(colData(vsd)) %>%
         cbind(cmdscale(vsd_sampleDistMatrix))
MDSplot_vst <- ggplot(mds_vst, aes(x = `1`, y = `2`,
                                shape = condition,
                                colour = sample)) +
                  geom_point(size = 3) +
                  theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
                  theme_classic() +
                  theme(panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
                  coord_fixed()
ggsave(MDSplot_vst,
       file = paste0(plotDir, "MDSplot_vst.pdf"), height = 20, width = 20, units = "cm")

# Poisson distance
mdsPois <- as.data.frame(colData(dds)) %>%
             cbind(cmdscale(samplePoisDistMatrix))
MDSplot_Pois <- ggplot(mdsPois, aes(x = `1`, y = `2`,
                                shape = condition,
                                colour = sample)) +
                  geom_point(size = 3) +
                  theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) +
                  theme_classic() +
                  theme(panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
                  coord_fixed()
ggsave(MDSplot_Pois,
       file = paste0(plotDir, "MDSplot_Pois.pdf"), height = 20, width = 20, units = "cm")

# Gene clustering
# Clustering of 20 genes with greatest rlog-transformed variance across samples
library(genefilter)
topVarGenes <- head(order(rowVars(assay(rld)), decreasing = T), 20)
mat <- assay(rld)[topVarGenes, ]
mat <- mat - rowMeans(mat)
anno <- as.data.frame(colData(rld)[, c("sample", "condition")])
pdf(paste0(plotDir, "gene_clustering_rld_topVar20.pdf"), height = 5, width = 7.5, onefile = F)
pheatmap(mat, annotation_col = anno)
dev.off()


