#!/bin/bash

i=$1

### Calculate alignment summary statistics for fastq files and for each bam file generated by the pipeline

# Output files with the suffix "fastq.stats" or "bam.stats" contain read counts,
# while output files with the suffix "alignments.stats" contain alignment counts
# This is to provide distinct alignment counts where reads map to multiple loci
gunzip -k ${i}.fastq.gz
cat ${i}.fastq | echo $((`wc -l`/4)) > ${i}.fastq.stats
samtools view ${i}_STAR.bam | cut -f 1 | sort | uniq | wc -l > ${i}_STAR.bam.stats
samtools view ${i}_STAR_mapped.bam | cut -f 1 | sort | uniq | wc -l > ${i}_STAR_mapped.bam.stats
samtools view ${i}_STAR_mapped_unique_sort.bam | cut -f 1 | sort | uniq | wc -l > ${i}_STAR_mapped_unique_sort.bam.stats
samtools view -c ${i}_STAR_mapped_multi.bam > ${i}_STAR_mapped_multi.bam.alignments.stats
samtools view -c ${i}_STAR_mapped_multi_primary_alignment.bam > ${i}_STAR_mapped_multi_primary_alignment.bam.alignments.stats
# Sanity check; (read) count generated by the following command should be the same as
# (alignment) count generated by the previous command
samtools view ${i}_STAR_mapped_multi_primary_alignment.bam | cut -f 1 | sort | uniq | wc -l > ${i}_STAR_mapped_multi_primary_alignment.bam.stats
samtools view ${i}_STAR_mapped_multi_primary_alignment_fq3_sort.bam | cut -f 1 | sort | uniq | wc -l > ${i}_STAR_mapped_multi_primary_alignment_fq3_sort.bam.stats
samtools view ${i}_STAR_mapped_both_sort.bam | cut -f 1 | sort | uniq | wc -l > ${i}_STAR_mapped_both_sort.bam.stats

paste -d "\t" ${i}.fastq.stats ${i}_STAR_mapped.bam.stats ${i}_STAR_mapped_unique_sort.bam.stats ${i}_STAR_mapped_multi.bam.alignments.stats ${i}_STAR_mapped_multi_primary_alignment.bam.alignments.stats ${i}_STAR_mapped_multi_primary_alignment.bam.stats ${i}_STAR_mapped_multi_primary_alignment_fq3_sort.bam.stats ${i}_STAR_mapped_both_sort.bam.stats > ${i}_alignment_summary_stats.txt
sed -i '1i Total_sequenced_reads\tAligned,_mismatches<=2\tUniquely_aligned\tMultiply_aligned_read_alignments\tPrimary_multiply_aligned_read_alignments\tPrimary_multiply_aligned\tPrimary_multiply_aligned,_MAPQ==3\tBoth' ${i}_alignment_summary_stats.txt
[ -d alignment_stats ] || mkdir alignment_stats
mv ${i}*.stats alignment_stats/
mv ${i}_alignment_summary_stats.txt alignment_stats/


### Remove no longer required fastq, text and sam files
rm ${i}.fastq

